### ISSUE TITLE
Integration with prescient?

### ISSUE THREAD

**Issue by HyunggyuJang** [ROCKTAKEY/ivy-migemo#1]
First of all, thanks for this assume package. Especially, I love that it chose to toggle rather than one shot to exploit `migemo`'s feature. 

The thing is, I was using `prescient` extension for `ivy`, and seems like it cannot work with your package, `ivy-migemo`. 

Appreciate your hard work! I'm looking forward your positive response, thanks!

**Comment by ROCKTAKEY** [comment]
Thanks for your issue! 
I added support for `ivy-prescient`. Please try `ivy-migemo-toggle-migemo` after MELPA update.

**Comment by HyunggyuJang** [comment]
Wow! Impressed by your quick response! I&rsquo;ve setup and adapted this package to my configuration, and tried out! However, it seems not work as expected.

Previously, I&rsquo;ve tweaked your package to reflect the `prescient` specific setting, something like this:

```elisp
(defadvice! +ivy-migemo-toggle-migemo ()
        :override #'ivy-migemo-toggle-migemo
        "Toggle the re builder to use/unuse migemo."
        (interactive)
        (setq ivy--old-re nil)
        (setq
         ivy--regex-function
         (pcase ivy--regex-function
           (`ivy--regex-fuzzy         #'ivy-migemo--regex-fuzzy)
           (`ivy--regex-plus          #'ivy-migemo--regex-plus)
           (`ivy-prescient-re-builder #'ivy-migemo--regex-plus)
           (`ivy-migemo--regex-fuzzy  #'ivy--regex-fuzzy)
           (`ivy-migemo--regex-plus   #'ivy-prescient-re-builder)
           (`swiper--re-builder
            (pcase (ivy-alist-setting ivy-re-builders-alist)
              (`ivy--regex-fuzzy         #'ivy-migemo--swiper-re-builder-migemo-regex-fuzzy)
              (`ivy--regex-plus          #'ivy-migemo--swiper-re-builder-migemo-regex-plus)
              (`+ivy-prescient-non-fuzzy #'ivy-migemo--swiper-re-builder-migemo-regex-plus)
              (`ivy-migemo--regex-fuzzy  #'ivy-migemo--swiper-re-builder-no-migemo-regex-fuzzy)
              (`ivy-migemo--regex-plus   #'ivy-migemo--swiper-re-builder-no-migemo-regex-plus)))
           (`ivy-migemo--swiper-re-builder-no-migemo-regex-fuzzy #'ivy-migemo--swiper-re-builder-migemo-regex-fuzzy)
           (`ivy-migemo--swiper-re-builder-no-migemo-regex-plus  #'ivy-migemo--swiper-re-builder-migemo-regex-plus)
           (`ivy-migemo--swiper-re-builder-migemo-regex-fuzzy    #'ivy-migemo--swiper-re-builder-no-migemo-regex-plus)
           (`ivy-migemo--swiper-re-builder-migemo-regex-plus     #'ivy-migemo--swiper-re-builder-no-migemo-prescient))))
```

Note that it was for the version before your refactoring, plus, I&rsquo;m using [Doom Emacs](https://github.com/hlissner/doom-emacs).

And with the current version, I become able to get rid of those impromptu hacks, and almost the next to the metal state, just bind `"M-m"` key to `ivy-migemo-toggle-migemo`, as you suggested.

However, it doesn&rsquo;t work; freezing the completion process. It looks like in some cases work, but even in that case, completion process became *extremely* slow.

Let me be more specific:

-   For `counsel` powered `find-file`, it doesn&rsquo;t work at all From the moment I activate the `ivy-migemo` function with the dedicated toggling keystroke, it literally stop filtering.
-   For `org-roam`&rsquo;s find file function, it works for some queries, but not for the others; not reliable at all.

I confirmed this kind of distorted behavior doesn&rsquo;t happen for the others, e.g. `ivy--regex-plus`, etc.

**Comment by HyunggyuJang** [comment]
For now, as workaround, using the following configuration:

```elisp
(use-package ivy-migemo
      :after ivy
      :init
      (define-key ivy-minibuffer-map (kbd "M-m") #'ivy-migemo-toggle-migemo)
      :general ([remap isearch-forward] #'swiper-isearch)
      :config
      (when (featurep 'ivy-prescient)
        (setf (alist-get 'ivy-prescient-re-builder ivy-migemo--regex-function-alist) #'+ivy-migemo-prescient)
        (add-to-list 'ivy-migemo--regex-function-alist
                  '(+ivy-migemo--swiper-re-builder-no-migemo-prescient . +ivy-migemo--swiper-re-builder-migemo-prescient))
        (add-to-list 'ivy-migemo--swiper-regex-function-alist
                  '(+ivy-migemo-prescient . +ivy-migemo--swiper-re-builder-migemo-prescient))
        (defalias '+ivy-migemo--swiper-re-builder-migemo-prescient #'ivy-migemo--swiper-re-builder-migemo-regex-plus)
        (defun +ivy-migemo--swiper-re-builder-no-migemo-prescient (str)
          "Apply `swiper--re-builder' forced to use `ivy-prescient-re-builder' with STR as argument."
          (ivy-migemo--swiper-re-builder-with str #'ivy-prescient-re-builder))
        (defalias '+ivy-migemo-prescient #'ivy-migemo--regex-plus)))
```

I&rsquo;ve changed it to be doom agnostic! Thanks for your consideration!!

**Comment by ROCKTAKEY** [comment]
Thanks for your information. I tried `counsel-find-file` and got the error:
```
Error in post-command-hook (ivy--queue-exhibit): (invalid-regexp "Regular expression too big")
```
This seems to mean slow process. 
Actually, I was wondering which is better,  creating prescient-specific re-builder as similar to prescient as possible, or  just aliasing `ivy-migemo--regex-plus`. I think the former is better **if it works**. But in this case, the latter is better. Not to work is evil. I will change it to use alias.

**Comment by ROCKTAKEY** [comment]
BTW, setting for swiper is needed? Prescient seems to be disabled on swiper (on `ivy-prescient-sort-commands`), so I don't apply prescient-specific re-builder to swiper.

**Comment by ROCKTAKEY** [comment]
I pushed alias version. Please try it if you don't mind.

**Comment by HyunggyuJang** [comment]
> BTW, setting for swiper is needed? Prescient seems to be disabled on swiper (on `ivy-prescient-sort-commands`), so I don&rsquo;t apply prescient-specific re-builder to swiper.

Hi there! Again, amazed by your activeness! I haven&rsquo;t checked it, and I agree with the decision made for `swiper` by the author of `precient` package. `swiper` seems to be nonsense to be sorted based on scored filtering.

Nonetheless, I needed to tweak this since [Doom Emacs](https://github.com/hlissner/doom-emacs)&rsquo;s `ivy` module set like this; so I tweaked. Come to think of it, maybe I can just drop the `prescient` specific setting for `swiper` as you suggested!

Thanks for sincere answer! I&rsquo;ll try your new patch in a moment!
